File: OpenGL/OpenGL/ConvexHull.cpp
#include<iostream>
#include<math.h>

using namespace std;
const int MAX_NUM = 50;
#define ERROR 0
/********************************************************/

/***A 3D Point Class***/
/********************************************************/

class Point3D
{
public:
	Point3D() {};
	Point3D(double x, double y, double z) :m_x(x), m_y(y), m_z(z) {};
	Point3D operator= (const Point3D & other);
	Point3D operator+ (const Point3D &other) const;
	Point3D operator- (const Point3D &other) const;
	//Cross product
	Point3D operator* (const Point3D &other) const;
	//Dot product
	double operator^ (const Point3D &other) const;
	~Point3D() {};

	double m_x;
	double m_y;
	double m_z;
};
Point3D Point3D::operator= (const Point3D &other)
{
	(*this).m_x = other.m_x;
	(*this).m_y = other.m_y;
	(*this).m_z = other.m_z;

	return *this;
}
Point3D Point3D::operator+ (const Point3D &other) const
{
	Point3D result(0, 0, 0);
	result.m_x = (*this).m_x + other.m_x;
	result.m_y = (*this).m_y + other.m_y;
	result.m_z = (*this).m_z + other.m_z;

	return result;
}
Point3D Point3D::operator- (const Point3D &other) const
{
	Point3D result(0, 0, 0);
	result.m_x = (*this).m_x - other.m_x;
	result.m_y = (*this).m_y - other.m_y;
	result.m_z = (*this).m_z - other.m_z;

	return result;
}
Point3D Point3D::operator* (const Point3D &other) const
{
	Point3D result(0, 0, 0);
	result.m_x = (*this).m_y*other.m_z - (*this).m_z * other.m_y;
	result.m_y = (*this).m_z * other.m_x - (*this).m_x*other.m_z;
	result.m_z = (*this).m_x*other.m_y - (*this).m_y * other.m_x;

	return result;
}
double Point3D::operator^ (const Point3D &other) const
{
	double result = 0;
	result = (*this).m_x*other.m_x +
		(*this).m_y*other.m_y + (*this).m_z * other.m_z;

	return result;
}

/********************************************************/

/***Plane Class***/
/********************************************************/
class Plane
{
public:
	bool is_hull;

	int v1;
	int v2;
	int v3;
	Plane() :v1(0), v2(0), v3(0), is_hull(false) {};
	Plane(int x, int y, int z, bool h) :v1(x), v2(y), v3(z), is_hull(h) {};
};
/********************************************************/

/***3D Convex Hull Class***/
/********************************************************/
class Convex3D
{
public:
	Convex3D() {};
	~Convex3D() {};
	bool IsVisibleToP(Point3D& p, Plane& f);
	void AddPointP(int p, int a, int b);
	bool CreateOriginTetrahedron();
	void ExtendConvexHull();

	int n; /*total number of real vertex*/
	Point3D points[MAX_NUM];/*Array contains all the points*/
	Plane triangleF[MAX_NUM]; /*All the triangle faces calculated before
								Including the faces that been deleted(is_hull = false)*/
	int orderedPlanes[MAX_NUM][MAX_NUM];
private:
	int triangleNum = 0;/*the total number of triangle faces on Convex hull*/

};

//We already made sure plane f is counterclock-reverse
bool Convex3D::IsVisibleToP(Point3D& p, Plane& f)
{
	Point3D vector1 = points[f.v2] - points[f.v1];
	Point3D vector2 = points[f.v3] - points[f.v1];
	Point3D vector3 = p - points[f.v1];
	return ((vector1*vector2) ^ vector3) > ERROR;
}
void Convex3D::AddPointP(int p, int a, int b)
{
	int face = orderedPlanes[a][b]; //face contains vector a->b
	if (triangleF[face].is_hull)
	{
		//if old hull plane is visible to point P, then need to delete the plane
		//and see its neighbors is also visible to P
		if (IsVisibleToP(points[p], triangleF[face]))
		{
			triangleF[triangleNum].is_hull = false;

			AddPointP(p, triangleF[face].v2, triangleF[face].v1);
			AddPointP(p, triangleF[face].v3, triangleF[face].v2);
			AddPointP(p, triangleF[face].v1, triangleF[face].v3);
		}
		//old hull plane is not visible to point P, we need P and a->b
		//to form new hull plane
		else
		{
			Plane newFace(b, a, p, true);
			orderedPlanes[p][b] = triangleNum;
			orderedPlanes[b][a] = triangleNum;
			orderedPlanes[a][p] = triangleNum;
			triangleF[triangleNum] = newFace;

			triangleNum += 1;
		}
	}
}
double Mod(Point3D a)
{
	return sqrt(a.m_x * a.m_x + a.m_y * a.m_y + a.m_z * a.m_z);
};
double Dist(Point3D a, Point3D b)
{
	Point3D vector = a - b;
	return Mod(vector);
}
double Area(Point3D a, Point3D b, Point3D c)
{
	Point3D vector1 = b - a;
	Point3D vector2 = c - a;
	double area = Mod(vector1 * vector2) / 2.0;
	return area;
}
double Volume(Point3D a, Point3D b, Point3D c, Point3D d)
{
	Point3D vector1 = b - a;
	Point3D vector2 = c - a;
	Point3D vector3 = d - a;
	double volume = abs(((vector1 * vector2) ^ vector3) / 6.0);
	return volume;
}
bool Convex3D::CreateOriginTetrahedron()
{
	bool success = false;
	cout << "HELLO" << endl;
	for (int i = 1; i < n; i++)
	{
		if (Dist(points[0], points[i]) > ERROR)
		{
			swap(points[1], points[i]);
			success = true;
			cout << "Found the point!!!" << endl;
			break;
		}
	}
	if (!success)
	{
		cout << "Same vertex" << endl;
		return success;
	}
	success = false;
	for (int i = 2; i < n; i++)
	{
		if (Area(points[0], points[1], points[i]) > ERROR)
		{
			swap(points[2], points[i]);
			success = true;
			cout << "Found the Edge!!!" << endl;
			break;
		}
	}
	if (!success)
	{
		cout << "Same edge" << endl;
		return success;
	}
	success = false;
	for (int i = 3; i < n; i++)
	{
		if (Volume(points[0], points[1], points[2], points[i]) > ERROR)
		{
			swap(points[3], points[i]);
			success = true;
			cout << "Found the face!!!" << endl;
			break;
		}
	}
	if (!success)
	{
		cout << "Same face" << endl;
		return success;
	}

	Plane p1(0, 1, 2, true);
	Plane p2(0, 1, 3, true);
	Plane p3(0, 2, 3, true);
	Plane p4(1, 2, 3, true);

	if (!IsVisibleToP(points[3], p1))
	{
		swap(p1.v2, p1.v3);
	}
	orderedPlanes[p1.v1][p1.v2] = triangleNum;
	orderedPlanes[p1.v2][p1.v3] = triangleNum;
	orderedPlanes[p1.v3][p1.v1] = triangleNum;
	triangleF[triangleNum] = p1;
	triangleNum += 1;

	if (!IsVisibleToP(points[2], p2))
	{
		swap(p2.v2, p2.v3);
	}
	orderedPlanes[p2.v1][p2.v2] = triangleNum;
	orderedPlanes[p2.v2][p2.v3] = triangleNum;
	orderedPlanes[p2.v3][p2.v1] = triangleNum;
	triangleF[triangleNum] = p2;
	triangleNum += 1;

	if (!IsVisibleToP(points[1], p3))
	{
		swap(p3.v2, p3.v3);
	}
	orderedPlanes[p3.v1][p3.v2] = triangleNum;
	orderedPlanes[p3.v2][p3.v3] = triangleNum;
	orderedPlanes[p3.v3][p3.v1] = triangleNum;
	triangleF[triangleNum] = p3;
	triangleNum += 1;

	if (!IsVisibleToP(points[0], p4))
	{
		swap(p1.v2, p1.v3);
	}
	orderedPlanes[p1.v1][p1.v2] = triangleNum;
	orderedPlanes[p1.v2][p1.v3] = triangleNum;
	orderedPlanes[p1.v3][p1.v1] = triangleNum;
	triangleF[triangleNum] = p4;
	triangleNum += 1;

	return success;
}
void Convex3D::ExtendConvexHull()
{
	for (int i = 4; i < n; i++)
	{
		for (int j = 0; j < triangleNum; j++)
		{
			if (IsVisibleToP(points[i], triangleF[j]))
			{

			}
		}
	}
}

int main()
{
	int n;
	Convex3D convexHull;
	cin >> n;
	convexHull.n = n;
	for (int i = 0; i < n; i++)
	{
		cin >> convexHull.points[i].m_x >> convexHull.points[i].m_y >> convexHull.points[i].m_z;
	}
	if (convexHull.CreateOriginTetrahedron())
	{
		for (int i = 0; i < 4; i++)
		{
			cout << convexHull.triangleF[i].v1 << " " << convexHull.triangleF[i].v2 << " " << convexHull.triangleF[i].v3 << endl;
		}

		for (int i = 0; i < 4; i++)
		{
			for (int j = 0; j < 4; j++)
			{
				cout << convexHull.orderedPlanes[i][j] << " ";
			}
			cout << endl;
		}
	}
	else
	{
		cout << "ERROE";
	}
	system("pause");
	return 0;
}

File: hw1/convexHull.py
# import open3d as o3d
# import numpy as np

# file_name = raw_input() or "test.txt"

# matrix = np.loadtxt(file_name, dtype = 'float')


# #pcl = o3dtut.get_bunny_mesh().sample_points_poisson_disk(number_of_points=2000)
# hull, _ = pcd.compute_convex_hull()
# hull_ls = o3d.geometry.LineSet.create_from_triangle_mesh(hull)
# hull_ls.paint_uniform_color((1, 0, 0))
# o3d.visualization.draw_geometries([pcd, hull_ls])

import numpy as np
import open3d as o3d
MAX_N = 2000
ERROR = 0
class Point3D:
    m_x = 0
    m_y = 0
    m_z = 0
    def __init__(self,x,y,z):
        self.m_x = x
        self.m_y = y
        self.m_z = z
    def DotProduct(self, other):
        return self.m_x * other.m_x + self.m_y * other.m_y + self.m_z * other.m_z
    def CrossProduct(self, other):
        coord_x = self.m_y * other.m_z - self.m_z * other.m_y
        coord_y = self.m_z * other.m_x - self.m_x * other.m_z
        coord_z = self.m_x * other.m_y - self.m_y * other.m_x
        result = Point3D(coord_x, coord_y, coord_z)
        return result
    def Minus(self,other):
        p = Point3D(0,0,0)
        p.m_x = self.m_x - other.m_x
        p.m_y = self.m_y - other.m_y
        p.m_z = self.m_z - other.m_z
        return p

class Face:
    is_hull = False
    v1 = 0
    v2 = 0
    v3 = 0
    def __init__(self, x, y, z, h):
        self.is_hull = h
        self.v1 = x
        self.v2 = y
        self.v3 = z

class Convex3D:
    n = 0
    points = []
    triangleF = []
    orderedFaces = np.zeros((MAX_N,MAX_N),dtype=int)
    triangleNum = 0

    def DirectedVolume(self, p, f):
        vector1 = Point3D.Minus(self.points[f.v2], self.points[f.v1])
        vector2 = Point3D.Minus(self.points[f.v3], self.points[f.v1])
        vector3 = Point3D.Minus(p, self.points[f.v1])
        return Point3D.DotProduct(Point3D.CrossProduct(vector1, vector2), vector3)

    def AddPointP(self, p,a,b):
        print("Current vertex: ",[self.points[p].m_x, self.points[p].m_y,self.points[p].m_z])
        face = self.orderedFaces[a][b]
        if self.triangleF[face].is_hull:
            print("Current Face: ", [self.points[self.triangleF[face].v1].m_x,self.points[self.triangleF[face].v1].m_y,self.points[self.triangleF[face].v1].m_z], 
                                    [self.points[self.triangleF[face].v2].m_x,self.points[self.triangleF[face].v2].m_y,self.points[self.triangleF[face].v2].m_z], 
                                    [self.points[self.triangleF[face].v3].m_x,self.points[self.triangleF[face].v3].m_y,self.points[self.triangleF[face].v3].m_z])
            print(self.DirectedVolume(self.points[p],self.triangleF[face]))
            if self.DirectedVolume(self.points[p],self.triangleF[face]) >  ERROR:
                print("Visible to P")
                self.ExtendVisiblePlane(p, face)
            else:
                newFace = Face(b,a,p,True)
                self.orderedFaces[p][b] = self.triangleNum
                self.orderedFaces[b][a] = self.triangleNum
                self.orderedFaces[a][p] = self.triangleNum
                
                self.triangleF.append(newFace)
                print("Add face: ", [self.points[newFace.v1].m_x,self.points[newFace.v1].m_y,self.points[newFace.v1].m_z], 
                                    [self.points[newFace.v2].m_x,self.points[newFace.v2].m_y,self.points[newFace.v2].m_z], 
                                    [self.points[newFace.v3].m_x,self.points[newFace.v3].m_y,self.points[newFace.v3].m_z])
                self.triangleNum += 1
    def ExtendVisiblePlane(self, p, tri_index):
        print("Ori Face: ",[self.points[self.triangleF[tri_index].v1].m_x,self.points[self.triangleF[tri_index].v1].m_y,self.points[self.triangleF[tri_index].v1].m_z],
                            [self.points[self.triangleF[tri_index].v2].m_x,self.points[self.triangleF[tri_index].v2].m_y,self.points[self.triangleF[tri_index].v2].m_z],
                            [self.points[self.triangleF[tri_index].v3].m_x,self.points[self.triangleF[tri_index].v3].m_y,self.points[self.triangleF[tri_index].v3].m_z])
        self.triangleF[tri_index].is_hull = False
        self.AddPointP(p,self.triangleF[tri_index].v2, self.triangleF[tri_index].v1)
        self.AddPointP(p,self.triangleF[tri_index].v3, self.triangleF[tri_index].v2)
        self.AddPointP(p,self.triangleF[tri_index].v1, self.triangleF[tri_index].v3)
    def CreateOriginTetrahedron(self):
        if len(self.points) < 4:
            print("Not enough points for 3D convexhull")
            return False
        success = False
        for i in range(1, self.n):
            if Dist(self.points[0], self.points[i]) > ERROR:
                self.points[1],self.points[i] = self.points[i], self.points[1]
                success = True
                print("Find two points")
                break
        if not success:
            print("Same vertex for all in list")
            return success
        success = False
        for i in range(2, self.n):
            if Area(self.points[0],self.points[1],self.points[i]) > ERROR:
                self.points[2], self.points[i] = self.points[i], self.points[2]
                success = True
                print("Find three points")
                break
        if not success:
            print("Same line for all in list")
            return success
        for i in range(3, self.n):
            if Volume(self.points[0], self.points[1], self.points[2], self.points[i]) > ERROR:
                self.points[3], self.points[i] = self.points[i], self.points[3]
                success = True
                print("Find Four Points")
                break
        if not success:
            print("Same plane for all in list")
            return success
        for i in range(4):
            face_tmp = Face((i+1)%4,(i+2)%4,(i+3)%4,True)
            if self.DirectedVolume(self.points[i], face_tmp) > 0:
                face_tmp.v2, face_tmp.v3 = face_tmp.v3, face_tmp.v2
            self.orderedFaces[face_tmp.v1][face_tmp.v2] = self.triangleNum
            self.orderedFaces[face_tmp.v2][face_tmp.v3] = self.triangleNum
            self.orderedFaces[face_tmp.v3][face_tmp.v1] = self.triangleNum
            self.triangleF.append(face_tmp)
            self.triangleNum += 1
        return success
    def ExtendConvexHull(self):
        for i in range(4,self.n):
            for j in range(self.triangleNum):
                if self.triangleF[j].is_hull and self.DirectedVolume(self.points[i], self.triangleF[j]) > ERROR:
                    self.ExtendVisiblePlane(i,j)
                    break
        for face in self.triangleF:
            if not face.is_hull:
                self.triangleF.remove(face)
def Mod(a):
    return np.sqrt(a.m_x * a.m_x + a.m_y * a.m_y + a.m_z * a.m_z)
def Dist(a, b):
    return Mod(Point3D.Minus(a,b))
def Area(a, b, c):
    vector1 = Point3D.Minus(b,a)
    vector2 = Point3D.Minus(c,a)
    return Mod(Point3D.CrossProduct(vector1, vector2)) / 2.0
def Volume(a, b, c, d):
    vector1 = Point3D.Minus(b, a)
    vector2 = Point3D.Minus(c, a)
    vector3 = Point3D.Minus(d, a)
    volume = abs(Point3D.DotProduct(Point3D.CrossProduct(vector1, vector2),vector3)) / 6.0
    return volume

if __name__ == "__main__":
    file_name = raw_input() or "test"
    matrix = np.loadtxt(file_name, dtype = 'float')
    convexHull = Convex3D()
    convexHull.n = np.size(matrix, axis = 0)
    # TODO: If the first row do not contain exactly one float, tell n is missing
    for i in range(convexHull.n):
        convexHull.points.append(Point3D(matrix[i][0],matrix[i][1],matrix[i][2]))
    if convexHull.CreateOriginTetrahedron():
        print("Extending ConvexHull")
        convexHull.ExtendConvexHull()
        for f in convexHull.triangleF:
            print("Calculated Faces: ",[convexHull.points[f.v1].m_x,convexHull.points[f.v1].m_y,convexHull.points[f.v1].m_z],
                            [convexHull.points[f.v2].m_x,convexHull.points[f.v2].m_y,convexHull.points[f.v2].m_z],
                            [convexHull.points[f.v3].m_x,convexHull.points[f.v3].m_y,convexHull.points[f.v3].m_z])

    tri_faces = []
    for f in convexHull.triangleF:
        tri_faces.append([f.v1, f.v2, f.v3])
    tri_faces = o3d.utility.Vector3iVector(np.array(tri_faces).reshape(-1,3))

    # tri_points_index = []
    # for f in convexHull.triangleF:
    #     if f.v1 not in tri_points_index:
    #         tri_points_index.append(f.v1)
    #     if f.v2 not in tri_points_index:
    #         tri_points_index.append(f.v2)
    #     if f.v3 not in tri_points_index:
    #         tri_points_index.append(f.v3)    
    # tri_points = []
    # for p in tri_points_index:
    #     coord = convexHull.points[p]
    #     tri_points.append([coord.m_x, coord.m_y, coord.m_z])
    # print(np.array(tri_points))

    # tri_points = o3d.utility.Vector3dVector(np.array(tri_points))
    tri_vertices = []
    for v in convexHull.points:
        tri_vertices.append([v.m_x, v.m_y, v.m_z])
    tri_vertices = o3d.utility.Vector3dVector(np.array(tri_vertices).reshape(-1,3))

    vertices = o3d.utility.Vector3dVector(matrix)
    pcd = o3d.geometry.PointCloud()
    pcd.points = vertices
    
    tri_mesh = o3d.geometry.TriangleMesh(tri_vertices, tri_faces)
    
    hull_ls = o3d.geometry.LineSet.create_from_triangle_mesh(tri_mesh)
    hull_ls.paint_uniform_color((1, 0, 0))
    mesh_frame = o3d.geometry.TriangleMesh.create_coordinate_frame(
    size=1, origin=[0, 0, 0])
    o3d.visualization.draw_geometries([pcd, hull_ls])
